File: Data Structures & Algorithms/count-connected-components/submission-7.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        
        adj = [ [] for i in range(n) ]
        visited = set()

        for u,v in edges:
            adj[u].append(v)
            adj[v].append(u)

        def dfs(node):

            # Base Case
            if node in visited:
                return 
            
            # Recursive Case
            visited.add(node)
            for nei in adj[node]:
                dfs(nei)
            return

        res = 0
        for node in range(n):
            if node not in visited:
                dfs(node)
                res += 1
        return res

        

File: Data Structures & Algorithms/binary-tree-diameter/submission-7.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    // Global variable;
    int maxDiameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {

        if ( root == null ) return 0;
        
        int diameter = maxHeight(root.left) + maxHeight(root.right);
        if ( diameter > maxDiameter) maxDiameter = diameter;

        // Traverse tree
        diameterOfBinaryTree(root.left);
        diameterOfBinaryTree(root.right);

        return maxDiameter;
    }

    private int maxHeight(TreeNode root) {

        // Base Case
        if ( root == null ) return 0;
        // Recursive Case
        return 1 + Math.max(maxHeight(root.left) , maxHeight(root.right));
    }
}






























// class Solution {

//     int maxDiameter = 0;

//     public int diameterOfBinaryTree(TreeNode root) {
//         dfs(root);
//         return maxDiameter;
//     }

//     private int dfs(TreeNode root) {

//         if ( root == null ) return 0;

//         int left  = dfs(root.left);
//         int right = dfs(root.right);
//         int diameter = left + right;

//         // Update maxDiameter
//         if ( diameter > maxDiameter) maxDiameter = diameter;

//         // Return height
//         return 1 + Math.max(left, right); 
//     }
      

// }
